staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/cbor.go

```diff
@@ -0,0 +1,224 @@
+	}
+	actual := tm.GetObjectKind().GroupVersionKind()
+	return &actual, nil
+}
+
+type Serializer struct {
+	metaFactory metaFactory
+	creater     runtime.ObjectCreater
+	typer       runtime.ObjectTyper
+}
+
+func NewSerializer(creater runtime.ObjectCreater, typer runtime.ObjectTyper) *Serializer {
+	return newSerializer(&defaultMetaFactory{}, creater, typer)
+}
+
+func newSerializer(metaFactory metaFactory, creater runtime.ObjectCreater, typer runtime.ObjectTyper) *Serializer {
+	return &Serializer{
+		metaFactory: metaFactory,
+		creater:     creater,
+		typer:       typer,
+	}
+}
+
+func (s *Serializer) Identifier() runtime.Identifier {
+	return "cbor"
+}
+
+func (s *Serializer) Encode(obj runtime.Object, w io.Writer) error {
+	// https://www.rfc-editor.org/rfc/rfc8949.html#name-self-described-cbor
+	if _, err := w.Write([]byte("\xd9\xd9\xf7")); err != nil {
+		return err
+	}
+
+	e := em.NewEncoder(w)
+	if u, ok := obj.(runtime.Unstructured); ok {
+		return e.Encode(u.UnstructuredContent())
+	}
+	return e.Encode(obj)
+}
+
+// gvkWithDefaults returns group kind and version defaulting from provided default
+func gvkWithDefaults(actual, defaultGVK schema.GroupVersionKind) schema.GroupVersionKind {
+	if len(actual.Kind) == 0 {
+		actual.Kind = defaultGVK.Kind
+	}
+	if len(actual.Version) == 0 && len(actual.Group) == 0 {
+		actual.Group = defaultGVK.Group
+		actual.Version = defaultGVK.Version
+	}
+	if len(actual.Version) == 0 && actual.Group == defaultGVK.Group {
+		actual.Version = defaultGVK.Version
+	}
+	return actual
+}
+
+func (s *Serializer) Decode(data []byte, gvk *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	actual, err := s.metaFactory.Interpret(data)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if gvk != nil {
+		*actual = gvkWithDefaults(*actual, *gvk)
+	}
+
+	if into != nil {
+		if u, ok := into.(runtime.Unstructured); ok {
+			dst := map[string]interface{}{}
+			if err := dm.Unmarshal(data, &dst); err != nil {
+				return nil, actual, err
+			}
+			// TODO: The UnstructuredList implementation of SetUnstructuredContent is
+			// not identical to what unstructuredJSONScheme does: (1) it retains the
+			// "items" key in its Object field, and (2) it does not infer a singular
+			// Kind from the list's Kind and populate omitted apiVersion/kind for all
+			// entries in Items.
+			u.SetUnstructuredContent(dst)
+
+			*actual = into.GetObjectKind().GroupVersionKind()
+			if len(actual.Kind) == 0 {
+				return nil, actual, runtime.NewMissingKindErr("<cbor>") // todo: diagnostic notation
+			}
+			if len(actual.Version) == 0 {
+				return nil, actual, runtime.NewMissingVersionErr("<cbor>") // todo: diagnostic notation
+			}
+			return into, actual, nil
+		}
+
+		types, _, err := s.typer.ObjectKinds(into)
+		switch {
+		case runtime.IsNotRegisteredError(err):
+			if err := dm.Unmarshal(data, into); err != nil {
```

Comment:
huh, so if the type is not Unstructured and is not registered to the scheme, we try to decode into it anyway?  Seems like we should be able to fail.